### PR TITLE
Port shoc_assumed_pdf functions individually (like in shoc.F90)

### PR DESCRIPTION
**Pull request by tcclevenger**
Closes https://github.com/E3SM-Project/scream/issues/1377. This is needed before all property tests can be switched to C++ from F90.

I didn't make individual BFB tests since I think I would need a work around for randomizing a single pack of data, and the BFBness is tested in `shoc_assumed_bfb`.

### CODE REVIEW

File: components/eamxx/src/physics/shoc/impl/shoc_assumed_pdf_compute_buoyancy_flux_impl.hpp [hunk 1]
```diff
@@ -0,0 +1,41 @@
...
+namespace scream {
+namespace shoc {
+
+/*
+ * Implementation of shoc_assumed_pdf_compute_buoyancy_flux. Clients should NOT
+ * #include this file, but include shoc_functions.hpp instead.
+ */
+
+template<typename S, typename D>
+KOKKOS_FUNCTION
```
Comment (bartgol) [inline components/eamxx/src/physics/shoc/impl/shoc_assumed_pdf_compute_buoyancy_flux_impl.hpp:17]: Given how tiny these functions are, we should probably use KOKKOS_INLINE_FUNCTION. Although the compiler will likely inline them anyways, we don't want to risk them to not be inlined.

**Comment by tcclevenger** [comment]
Investigating CPU fail for baseline tests.

**Comment by tcclevenger** [comment]
@bartgol We get compile time warning about inlining functions without definition on CPU which I believe is because we are interfacing through `eti/func_name.cpp`, so I need to revert.

**Comment by bartgol** [comment]
> @bartgol We get compile time warning about inlining functions without definition on CPU which I believe is because we are interfacing through `eti/func_name.cpp`, so I need to revert.

You mean that there's no available definition, and the code is linked? How does that even work on GPU then, workout the relocatable device code flags?

**Comment by tcclevenger** [comment]
> You mean that there's no available definition, and the code is linked? How does that even work on GPU then, workout the relocatable device code flags?

In `shoc_functions.hpp`, after all the function declarations, we have 
```
#ifdef ENABLE_GPU and not RELOCATABLE_DEVICE
  #include func_name_impl.hpp
#endif
```

**Comment by bartgol** [comment]
Oh, I see. Our ETI strategy is conflicting with the inline declaration. That makes sense. I wonder if ETI makes sense for such small functions...

**Comment by tcclevenger** [comment]
ETI probably doesn't make sense for these. Considered just adding the definitions to the .hpp declaration, but wouldn't want to clutter it up.

**Comment by bartgol** [comment]
> ETI probably doesn't make sense for these. Considered just adding the definitions to the .hpp declaration, but wouldn't want to clutter it up.

You can simply include the impl files, but, unlike other functions, you don't guard with any ifdef, so they always get included. Then, you can mark them inline.

**Review by bartgol — APPROVED** [review]

**Pull request merged**